pkg/scheduler/cache/cache.go

```diff
@@ -176,6 +184,251 @@ type imageState struct {
+func NewRegexMatcher(regex string) (*MemberMatcher, error) {
+	_, err := regexp.Compile(regex)
+	if err != nil {
+		klog.Errorf("Failed to compile regex %s: %v", regex, err)
+		return nil, err
+	}
+	return &MemberMatcher{Regex: regex}, nil
+}
+
+// HyperNodeTreeNode reprensent a single hypernode in the tree
+type HyperNodeTreeNode struct {
+	Name           string
+	Tier           int
+	Parent         *HyperNodeTreeNode
+	Members        sets.Set[*HyperNodeTreeNode]
+	MemberMatchers sets.Set[*MemberMatcher]
+	Nodes          sets.Set[*schedulingapi.NodeInfo]
+	NodeMatchers   sets.Set[*MemberMatcher]
+	HyperNodeRef   *networktopov1alpha1.HyperNode
+}
+
+func NewHyperNodeTreeNode(name string) *HyperNodeTreeNode {
+	return &HyperNodeTreeNode{
+		Name:           name,
+		Tier:           -1, // Initialize to an invalid value
+		Members:        sets.New[*HyperNodeTreeNode](),
+		MemberMatchers: sets.New[*MemberMatcher](),
+		Nodes:          sets.New[*schedulingapi.NodeInfo](),
+		NodeMatchers:   sets.New[*MemberMatcher](),
+	}
+}
+
+type HyperNodeSet sets.Set[*HyperNodeTreeNode]
+
+// HyperNodeForestBuilder helps in building the forest of hypernodes
+type HyperNodeForestBuilder struct {
+	nodes map[string]*HyperNodeTreeNode // Map to store nodes temporarily
+	roots sets.Set[string]              // Set of root node names
+}
+
+// NewHyperNodeForestBuilder initializes and returns a ForestBuilder
+func NewHyperNodeForestBuilder() *HyperNodeForestBuilder {
+	return &HyperNodeForestBuilder{
+		nodes: make(map[string]*HyperNodeTreeNode),
+		roots: sets.New[string](),
+	}
+}
+
+// insertHyperNodeMember inserts a member into a hypernode and updates the forest
+func insertHyperNodeMember(hntn *HyperNodeTreeNode, hnfb *HyperNodeForestBuilder, memberName string, memberNode *HyperNodeTreeNode, sc *SchedulerCache) {
+	hntn.Members.Insert(memberNode)
+	memberNode.Parent = hntn
+	if hnfb.roots.Has(memberName) {
+		hnfb.roots.Delete(memberName)
+	}
+	sc.HyperNodeListByTier[memberNode.Tier].Insert(memberName)
+
+	sc.HyperNodes[hntn.Name].Union(sc.HyperNodes[memberName])
+}
+
+// addHyperNode adds a HyperNodeTreeNode to the forest based on v1apha1.HyperNode
+// Adding a hypernode can result in updating the entire forest, as it may add new
+// nodes, or link existing nodes - which could demote nodes from roots to child.
+// Because of regex matching, this can trigger ripple effects in the forest.
+func (sc *SchedulerCache) addHyperNode(hn *networktopov1alpha1.HyperNode) {
+	hnfb := sc.hypernodeForestbuilder
+	hntn := NewHyperNodeTreeNode(hn.Name)
+	hntn.HyperNodeRef = hn
+	hnfb.nodes[hn.Name] = hntn
+
+	tierStr := hn.Spec.Tier
+	tier, err := strconv.Atoi(tierStr)
+	if err != nil {
+		klog.Errorf("Failed to convert tier %s to int: %v", tierStr, err)
+		return
+	}
+	hntn.Tier = tier
+	klog.V(4).Infof("Adding hypernode %s to the forest, in %d tier", hn.Name, tier)
+
+	// Iterate over members and add them to the forest
+	for _, member := range hn.Spec.Members {
```

Comment:
This can be wrapped as a func, it' the logic to build a hyperNodeTreeNode, and the last thing to do is update the `sc.HyperNodeListByTier` and `sc.HyperNodes`. 
And can we just use one var to express the tree?